### ISSUE TITLE
new check: any unused metadata

### ISSUE THREAD

**Issue by trel** [irods/irods_client_zone_management_tool#108]
is there any unused metadata in the catalog that can be removed (`iadmin rum`)

**Comment by ganning127** [comment]
Is there a way to show unused metadata? Or would there just be a button that does the equivalent of `iadmin rum` (ie removes unused metadata) through the `/admin) endpoint?

**Comment by korydraughn** [comment]
I doubt there's a command or query you can run that shows this information.

I see two ways of approaching this.
- Approach A _(performance is guaranteed to be bad for large datasets)_
    1. Find all metadata attached to users, resources, data objects, and collections
    2. Gather all metadata in the system
    3. Subtract the metadata in (1) from (2)
- Approach B _(significantly better solution )_
    - Define and use a specific query that can find this information in one go

**Comment by trel** [comment]
or... we implement a 'rum' dry-run?  that just reports what *would* happen?

**Comment by trel** [comment]
i suppose that's kind of approach B

**Comment by ganning127** [comment]
I think the `iadmin rum` dry-run would be the most "consistent". If there was a different query created that finds unused metadata, it might malfunction in edge cases that `iadmin rum` performs fine on.

Or, if `iadmin rum` is using a query behind the scenes, if you can provide me with the query, I can just remove the "deletion" part of that command to be a "dry-run"

**Comment by trel** [comment]
This is the function...

https://github.com/irods/irods/blob/main/plugins/database/src/db_plugin.cpp#L476

**Comment by ganning127** [comment]
Looking at that file, there are 3 different delete operations depending on the type of database, I believe (Oracle, MySQL, Postgres). Since I think I'm running Postgres, I've been trying to select and grab unused metadata, but I'm not really sure how to do this. 

I created a data object and added some metadata to it using `imeta`. Next, I deleted that data object, and verified that the metadata was still there using `iquest "SELECT META_DATA_ATTR_ID"`. However, when I used the SQL query for Postres (from the removeAVUs function), but replacing `DELETE` with `SELECT *`: `SELECT * FROM R_META_MAIN where meta_id in (select meta_id from R_META_MAIN except select meta_id from R_OBJT_METAMAP);` but got 0 rows. I also tried the queries for the other databases, but they either returned 0 rows (for MySQL), or errored out (for Oracle).

**Comment by trel** [comment]
The subselect is doing the 'finding' of what should be 'deleted'... so just running the subselect, I think, will be enough to see if there is anything that qualifies...  and we just want to see how many... so...

I think...
```
select count(meta_id) from R_META_MAIN except select meta_id from R_OBJT_METAMAP
```

**Comment by ganning127** [comment]
Hmmm, that command isn't working for me (or I'm not creating the un-used metadata correctly). While I figure that out, if there is a specific query that can be defined, I assume that we're going to take the same approach as with the users with invalid zones issue?

So, if the specific query is defined, we'll run that and show how many pieces of un-used metadata there are. If the specific query isn't defined, we'll use the inefficient method and ask the user to define the specific query?

**Comment by trel** [comment]
Ah, can't use except with count(meta_id) compared to the set coming from the subselect (just meta_id)...

instead... make it a subquery itself, using an alias, and then all wrapped in a count()...
```
select count(*) from (select meta_id from R_META_MAIN except select meta_id from R_OBJT_METAMAP) as unused;
```

**Comment by ganning127** [comment]
Yep, was just on a call with Kory and we figured out that the needed query to find the number of unused metadata was:

```
SELECT count(t.*) FROM (select meta_id from R_META_MAIN except select meta_id from R_OBJT_METAMAP) t;
```


Also, I think the approach will be:
1. If the user has the specific query above defined, we'll run that query and return the count
2. If the user doesn't have the specific query above defined, we'll need to:
- Figure out all metadata associated with data objects (`SELECT META_DATA_ATTR_ID, DATA_ID`), users (`SELECT_META_USER_ATTR_ID`, `USER_ID`), resources (`SELECT META_RESC_ATTR_ID, RESC_ID`), and collections (`SELECT META_COLL_ATTR_ID, COLL_ID`), and combine these metadata ids into a single list of all "used metadata".
- Figure out all the metadata in the system using (`SELECT META_DATA_ATTR_ID`).

Since we're only interested in the length, we can take the difference in lengths of the two lists. If there is unused metadata, show an error (and ask them to add the specific query). If not, show a warning that asks them to add the specific query.

**Comment by trel** [comment]
getting the full lengths may be very heavyweight.  might be three levels...
1) specific query
2) check for 'not too big', if so, just complain and ask nicely to add the specific query
3) if not too big, do the inefficient math and show the count, AND ask to add the specific query

**Comment by ganning127** [comment]
PR merged, closing as completed
